[BEAM-3659] Port UserScoreTest off DoFnTester

cademarkegard opened the pull request:
Ports UserScoreTest off the `DoFnTester` in favor of testing through a real pipeline. This is subtask of the story of [BEAM-3650](https://issues.apache.org/jira/browse/BEAM-3650) Deprecating and removing the DoFnTester.

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/) [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/)
Python | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | --- | ---

aromanenko-dev commented on examples/java/src/test/java/org/apache/beam/examples/complete/game/UserScoreTest.java:
```diff
@@ -92,13 +104,12 @@
   /** Test the {@link ParseEventFn} {@link org.apache.beam.sdk.transforms.DoFn}. */
   @Test
```
I'd add `@Category(ValidatesRunner.class)` here

cademarkegard replied:
added

kennknowles replied:
Actually I have to issue a different comment. I didn't notice this. We save {{ValidatesRunner}} for specification tests of the Beam core. For examples, the `TestPipeline` that you use will already pick up whatever runner is configured. So you don't have to do anything for this to work as intended, running the pipeline on whatever runner is set up.

cademarkegard replied:
thanks for the explanation, I've removed it

aromanenko-dev replied:
Yes, agree with @kennknowles , sorry for misleading with that. I followed the same strategy as we have for other tests (in this test suite and in other similars) that are marked with this annotation. Don't we have to clean them up?

In the same time, `ValidatesRunner extends NeedsRunner` and Javadoc of `NeedsRunner` says:
```
/**
 * Category tag for validation tests which utilize {@link TestPipeline} for execution and expect to
 * be executed by a {@link PipelineRunner}.
 */
```
So, don't we have to mark all "TestPipeline"-ed tests with `NeedsRunner`?

kennknowles replied:
That's right. You do need to put the DirectRunner on the classpath and TestPipeline will pick it up.

The `@Category(NeedsRunner.class)` annotation is pretty much just so we can scrape them out of the SDK jar, since it cannot depend on the DirectRunner. In maven it would be a (spurious) cyclic dependency problem. With gradle I think not a problem, actually.

aromanenko-dev commented on examples/java/src/test/java/org/apache/beam/examples/complete/game/UserScoreTest.java:
```diff
@@ -92,13 +104,12 @@
   @Test
   public void testParseEventFn() throws Exception {
-    DoFnTester<String, GameActionInfo> parseEventFn = DoFnTester.of(new ParseEventFn());
+    PCollection<String> input = p.apply(Create.of(GAME_EVENTS).withCoder(StringUtf8Coder.of()));
```
I don't think we need to add explicitly default coder `StringUtf8Coder`

cademarkegard replied:
I originally had it because the other tests were also using it. I've removed the coder from the other examples to stay consistent.

kennknowles commented on examples/java/src/main/java/org/apache/beam/examples/complete/game/UserScore.java:
```diff
@@ -108,8 +112,39 @@ public String getKey(String keyname) {
+    }
+
+    @Override
+    public int hashCode() {
```
Perhaps `Objects.hash(...)`?

cademarkegard replied:
👍

kennknowles commented on examples/java/src/main/java/org/apache/beam/examples/complete/game/UserScore.java:
```diff
@@ -108,8 +112,39 @@ public String getKey(String keyname) {
+      if (this == o) {
+        return true;
+      }
+      if (o == null || getClass() != o.getClass()) {
```
There is a subtle thing here. If you want to insist that the class is exactly this class, then you should use `o.getClass() == GameActionInfo.class`. That matches your use of field references below. Otherwise, you should change each `this.team` into `this.getTeam()`, and so on for score, user, etc. Basically, embrace inheritance or prevent inheritance. FWIW inheritance and equivalence relations are opposed anyhow.

cademarkegard replied:
Great comment. I liked the way you laid out the options, it was a nice way for me to learn about the subtleties. Thanks! I ended up choosing it `o.getClass() == GameActionInfo.class` option so it is explicit we're checking this specific class.

cademarkegard replied:
nevermind, I had to revert it, FindBugs reported it:

```EQ_GETCLASS_AND_CLASS_CONSTANT: equals method fails for subtypes
This class has an equals method that will be broken if it is inherited by subclasses. It compares a class literal with the class of the argument (e.g., in class Foo it might check if Foo.class == o.getClass()). It is better to check if this.getClass() == o.getClass().``` 

So I changed it from the fields to the getter methods.

kennknowles replied:
Findbugs and I have a mild feud over this one, as I care more about the usual properties of equality than the usual properties of inheritance. One is much older and more clearly timeless than the other... :-)

Your change is g2g though. This is a corner case with no right answer, so making findbugs happy by calling a getter is just fine with me.

cademarkegard replied:
hah, thanks!

cademarkegard commented:
@kennknowles @aromanenko-dev I've addressed the comments, however, it seems that the java pre-commit is failing due to some GRPC test. Is this a known issue? Should I just trigger another build?

aromanenko-dev commented:
Run Java PreCommit

cademarkegard commented:
Run Java PreCommit

kennknowles commented:
I'm not sure that pushes to `master` change the tested commit pointed to by the ref `pull/7126/merge`. You might just rebase and force push to ensure you are testing against the latest master where maybe flakes are better.

cademarkegard commented:
thanks @kennknowles that worked. 🎉 

@aromanenko-dev could you please review again?

aromanenko-dev commented:
I'm merging since, finally, the changes were approved.
